### PR TITLE
[WILL-NOT-MERGE] Rework network layer by Netty

### PR DESCRIPTION
**Pull request by pan3793**
## What changes in this PR?

Generally, this PR is aim to drop blocking socket and migrate network layer to Netty to achieve better performance and possibility to provide a Async Client API.

Currently implementation should be considered a prototype:
- [x] Rewrite network layer by Netty
- [x] Pass build 
- [x] Pass Unit Tests
- [x] Pass Integration Tests
- [x] Async Client API
- [ ] `ClickHouseSQLException` should only use in JDBC interface
- [ ] Avoid `ByteBuf` leak[1]
- [ ] Support codec compressed `Block`
- [ ] Elegantly handle half-response packet[2]
- [ ] Better performance

[1] Netty `ByteBuf` use reference count to track `ByteBuf` lifecycle for GC-free when use off-heap allocator, and it provide a `ResourceLeakDetector` to detect leaked `ByteBuf` on runtime, which "leaked" means the `ByteBuf` has no referenced and can be GC, but it's reference counter > 0. It's no affect on heap `ByteBuf`.
[2] We have a offline discussion with @sundy-li @zhang2014 @zhicwu, and the solution is move decode process to a dedicated thread pool using blocking IO API rather than decode in Netty eventloop.

### CODE REVIEW

**Comment by zhou-hao** [comment]
why close the PR?

**Comment by pan3793** [comment]
> why close the PR?

@zhou-hao Thanks for watching this PR. At the begining, I believe that Netty is the fastest network framework in Java world, so started this refactor, but it's not a fluky thing, I went back and forth and final decided to abandon it.

TL;DR The most important reason is I didn't get performance improvement but downgrade with Netty.

I will update the detail if I have time later.

**Comment by zhou-hao** [comment]
感谢回复，看上去你是个中国人？那就中文交流？
我认为，netty并不能提升RPS，但是当单个操作耗时比较久时，是否更有优势？

**Comment by pan3793** [comment]
@zhou-hao There may some international friends watching this issue, we can have a discussion offline in email or wechat if you are interested in it.

There is an important thing in network program (specific TCP based protocol here) is how to identify a packet(or frame)?

Typically, the protocol define a DELIMITER, or put the frame length at the frame header, but ClickHouse native protocol has none of them.
When I use netty to receive the packet, I didn't find a fast approach to detect is there a complete packet in buffer, so I choose the a simple way, try decode packet every time on receive the bytes, and cache the bytes if decode failed.

After second thought, I found a way to handle this issue, to build a huge state machine to track the decode state of packet, but it will cost lots of time, I didn't try it.

**Comment by zhou-hao** [comment]
Thinks!!!

**Comment by pan3793** [comment]
Another thing is that `ByteBuf` use reference counter to manage the lifecycle. After switching all `byte[]` to `ByteBuf`, user must ensure that close all `Connection`, `Statement`, `ResultSet` they created, otherwise, netty will complain memory leak (there is no leak if we use heap allocator, but off-heap does).

Even handle `ByteBuf` lifecycle properly, through the cpu profiler, I saw the performance downgrade, and a lot of time cost on `ByteBuf` allocate and recycle.

I'm not expert in memory management, maybe I don't use `ByteBuf` in properly, and I'm open to discuss/learn how to manage memory efficiently.

**Comment by pan3793** [comment]
> 我认为，netty并不能提升RPS，但是当单个操作耗时比较久时，是否更有优势？

Yes, NIO style does. But I don't think ClickHouse is designed for such scenario.

**Comment by zhou-hao** [comment]
How about using [reactor netty](https://github.com/reactor/reactor-netty) to implement ClickHouse Client?

**Comment by pan3793** [comment]
> How about using [reactor netty](https://github.com/reactor/reactor-netty) to implement ClickHouse Client?

Not familiar with this project, but as my understanding, it didn't solve the problems I list above.

ClickHouse support gRPC from v21.1.2.15-stable 2021-01-18, link: https://github.com/ClickHouse/ClickHouse/blob/master/CHANGELOG.md#clickhouse-release-v211215-stable-2021-01-18

I think it's a right direction of high performance. I'm more interesting in integration on ClickHouse with Spark now. I can't promise a roadmap/plan for this project, because I'm not salaried for it.
